subiquity/controllers/installprogress.py

```diff
@@ -40,10 +49,109 @@ class InstallState:
     ERROR = -1
 
 
+class WaitForCurtinEventsTask(BackgroundTask):
+
+    def __init__(self, controller):
+        self.controller = controller
+        self.waited = 0.0
+
+    def start(self):
+        pass
+
+    def _bg_run(self):
+        while self.controller._event_indent and self.waited < 5.0:
+            time.sleep(0.1)
+            self.waited += 0.1
+        log.debug("waited %s seconds for events to drain", self.waited)
+
+    def end(self, observer, fut):
+        # Will raise if command failed:
+        fut.result()
+        self.controller._install_event_start("installing snaps")
+        observer.task_succeeded()
+
+
+class DownloadSnapTask(BackgroundTask):
+
+    def __init__(self, controller, download_dir, snap_name, channel):
+        self.controller = controller
+        self.download_dir = os.path.join(download_dir, snap_name)
+        self.snap_name = snap_name
+        self.channel = channel
+
+    def start(self):
+        self.controller._install_event_start(_("downloading {}").format(self.snap_name))
+        os.mkdir(self.download_dir)
```

Comment:
This may not be sufficient if download_dir or snap_name includes path separators.    os.makedirs()

Reply:
Ah so this code makes some assumptions about the directory layout we're working in and I want it to fail if the assumptions fail. I should explain all this a bunch more though... added a long comment